core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java

```diff
@@ -0,0 +1,336 @@
+        state.nodes(discoBuilder);
+        state.metaData(MetaData.builder().put(indexMetaData, false).generateClusterUuidIfNeeded());
+        return state.build();
+    }
+
+    @Test
+    public void testLocalOperationWithoutBlocks() throws ExecutionException, InterruptedException {
+        final boolean masterOperationFailure = randomBoolean();
+
+        Request request = new Request();
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+
+        final Throwable exception = new Throwable();
+        final Response response = new Response();
+
+        clusterService.setState(mockedClusterState(MASTER_EXISTS, LOCAL_NODE_IS_MASTER));
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected void masterOperation(Request request, ClusterState state, ActionListener<Response> listener) throws Exception {
+                if (masterOperationFailure) {
+                    listener.onFailure(exception);
+                } else {
+                    listener.onResponse(response);
+                }
+            }
+        }.execute(request, listener);
+
+        if (masterOperationFailure) {
+            try {
+                listener.get();
+                fail("Expected block but returned proper result");
+            } catch (ExecutionException ex) {
+                assertThat(ex.getCause(), equalTo(exception));
+            }
+        } else {
+            assertThat(listener.get(), equalTo(response));
+        }
+    }
+
+    @Test
+    public void testLocalOperationWithBlocks() throws ExecutionException, InterruptedException {
+        final boolean retryableBlock = randomBoolean();
+        final boolean unblockBeforeTimeout = randomBoolean();
+
+        Request request = new Request().masterNodeTimeout(TimeValue.timeValueSeconds(unblockBeforeTimeout ? 60 : 1));
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+
+        ClusterBlock block = new ClusterBlock(1, "", retryableBlock, true,
+                randomFrom(RestStatus.values()), ClusterBlockLevel.ALL);
+        ClusterState stateWithBlock = ClusterState.builder(mockedClusterState(MASTER_EXISTS, LOCAL_NODE_IS_MASTER))
+                .blocks(ClusterBlocks.builder().addGlobalBlock(block)).build();
+        clusterService.setState(stateWithBlock);
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+                Set<ClusterBlock> blocks = state.blocks().global();
+                return blocks.isEmpty() ? null : new ClusterBlockException(blocks);
+            }
+        }.execute(request, listener);
+
+        if (retryableBlock && unblockBeforeTimeout) {
+            Thread.sleep(TimeValue.timeValueSeconds(1).millis());
```

Comment:
what does the sleep here do? should the `setState` in next line not be sufficient to make the listener return or do you want to test that the timeout works here? Can we also add an `assertFalse(listener.isDone())` ?